lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs

```diff
@@ -103,18 +110,41 @@ public byte[] Call(byte[] memento)
                     GC.WaitForPendingFinalizers();
                 }
 
-                using (
-                    MapInputWithControlMessage<TMapInput> mapInput = _dataAndMessageReceiver.Receive())
+                try
                 {
-                    controlMessage = mapInput.ControlMessage;
-                    if (controlMessage != MapControlMessage.Stop)
+                    using (
+                    MapInputWithControlMessage<TMapInput> mapInput = _dataAndMessageReceiver.Receive(_cancellationSource))
                     {
-                        _dataReducer.Send(_mapTask.Map(mapInput.Message));
+                        controlMessage = mapInput.ControlMessage;
+                        if (controlMessage != MapControlMessage.Stop)
+                        {
+                            _dataReducer.Send(_mapTask.Map(mapInput.Message), _cancellationSource);
+                        }
                     }
                 }
+                catch (OperationCanceledException e)
+                {
+                    Logger.Log(Level.Warning, "Received OperationCanceledException in MapTaskHost with message: {0}.", e.Message);
+                    break;
+                }
+                catch (IOException e)
+                {
+                    Logger.Log(Level.Error, "Received IOException in MapTaskHost with message: {0}.", e.Message);
+                    if (!_cancellationSource.IsCancellationRequested)
+                    {
+                        throw new IMRUTaskGroupCommunicationException(TaskManager.TaskGroupCommunicationError);
+                    }
+                    break;
+                }
+                catch (Exception e)
+                {
+                    Logger.Log(Level.Error, "Received Exception in MapTaskHos with exception type {0} and stack trace {1}.", e.GetType(), e.StackTrace);
```

Comment:
May be eventually we can remove it. For now, I just want to see all he possible exceptions that could be caught in IMRUTasks.